worldedit-core/src/main/java/com/sk89q/worldedit/util/SideEffectApplier.java

```diff
@@ -0,0 +1,93 @@
+/*
+ * WorldEdit, a Minecraft world manipulation toolkit
+ * Copyright (C) sk89q <http://www.sk89q.com>
+ * Copyright (C) WorldEdit team and contributors
+ *
+ * This program is free software: you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as published by the
+ * Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License
+ * for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package com.sk89q.worldedit.util;
+
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Maps;
+
+import java.util.Arrays;
+import java.util.EnumMap;
+import java.util.Map;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+public class SideEffectApplier {
+    private static final SideEffectApplier DEFAULT = new SideEffectApplier();
+    private static final SideEffectApplier NONE = new SideEffectApplier(
+            Arrays.stream(SideEffect.values()).collect(Collectors.toMap(Function.identity(), SideEffect::getDefaultValue))
+    );
+
+    private final Map<SideEffect, SideEffect.State> sideEffects;
+    private boolean requiresCleanup = false;
+    private boolean appliesAny = false;
+
+    private SideEffectApplier() {
+        this.sideEffects = ImmutableMap.of();
+        updateSideEffects();
+    }
+
+    public SideEffectApplier(Map<SideEffect, SideEffect.State> sideEffects) {
+        this.sideEffects = Maps.immutableEnumMap(sideEffects);
+        updateSideEffects();
+    }
+
+    private void updateSideEffects() {
```

Comment:
I find it weird that you have a separate method for this, rather than dumping it in the `(Map)` constructor, and using `this(ImmutableMap.of())` for the no-arg constructor (`Maps.immutableEnumMap` can handle empty maps fine)